Enable Aarch64 support via SIMDe.

pgwipeout opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
This pull request enables compiling on Aarch64, while retaining full previous compatibility.
I still need someone who can test against PowerPC.

To do this, I've added SIMDe as a dependency, which provides an audited universal SIMD abstraction layer.
This provides full performance for the native intrinsics, while allowing relatively fast translation for aarch64.
SIMDe also includes Hedley, an acceleration layer that can improve performance of existing code.
I've converted all currently used x86 intrinsics to use the SIMDe layer.
I've enabled automatic aarch64 detection in cmake.
**EDIT** 
SIMDe is no longer a dependency, I took a snapshot of the files actually used and put them in libobs/util/aarch.
I've written an abstraction layer to translate between the native intrinsics and SIMDe.
This abstraction layer is only used when aarch64 is detected, otherwise everything stays as normal.

Should additional intrinsics that I haven't included in the abstraction layer be added to the project, I will have to manually add them here as well.


### Motivation and Context
Aarch64 hardware has reached a point where it is viable as a streaming platform.
By allowing OBS-Studio to compile on aarch64 platforms, we open it up to use on the most common computing platform in the world.
This can be used as a platform for enabling new features and more robust hardware accelerated video encoding.
Further down the line, I hope to have a plugin to allow OBS-Studio to use Gstreamer as a hardware acceleration platform.
Since Gstreamer is starting to support Windows, and is relatively common on Linux, this would also provide a fallback method for x86 streaming.
**EDIT**
There is prototype obs-gstreamer code that provides hardware acceleration in the dev branch.
It is still under development and the hardware acceleration code needs work to function on specific boards.

### How Has This Been Tested?
This has been compile tested on the rk3328 platform.
This has been compile and run tested on the Jetson Nano platform running the Linux4Tegra OS (Ubuntu 18.03 with nVidia binaries).
This is not compatible yet with nVidia Tegra running nouveau, due to it lacking any functional hardware decoding.
Compiles successfully on x86/x64.
**Needed** Test on PowerPC.

### Types of changes
New feature (non-breaking change which adds functionality)
Supports aarch64.

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

RytoEX commented:
A few notes on commit messages:
* prefixes are generally lowercase except in some instances (CI, UI)
* commit message subjects don't end in periods

pgwipeout commented:
Should be ready for inclusion.

kkartaltepe commented on deps/media-playback/CMakeLists.txt:
```diff
@@ -28,6 +28,9 @@ if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64le")
 			PUBLIC
 			-mvsx)
 	add_compile_definitions(NO_WARN_X86_INTRINSICS)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
```
Just make the check below appropriate instead of adding this which does nothing and depends on the order of the if statements.

pgwipeout replied:
This is absolutely necessary, as otherwise the `elseif(NOT MSVC)` immediately following applies the -mmmx, -msse, and -msse2 flags which break portability. (Also, depending on the compiler these may not even be necessary)

Now if you are asking me to clean up the entire selection process, that's fine too, but please be specific.

kkartaltepe replied:
indeed the "check below" im referring to is the MSVC check. As I see it we shouldnt blindly apply these extensions if there is already an architecture check available to us. This also makes support aarch64 simpler since it will do nothing.

kkartaltepe commented on libobs/CMakeLists.txt:
```diff
@@ -470,6 +482,9 @@ if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64le")
 		PUBLIC
 			-mvsx)
 	add_compile_definitions(NO_WARN_X86_INTRINSICS)
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
```
Same as above.

kkartaltepe commented on libobs/util/aarch/check.h:
```diff
@@ -0,0 +1,258 @@
+/* Check (assertions)
+ * Portable Snippets - https://gitub.com/nemequ/portable-snippets
```
Typo

kkartaltepe commented:
Im not sure if all this code really belongs in `libobs/util`, since its mostly 
https://github.com/nemequ/simde &
https://github.com/nemequ/portable-snippets
shredded to fit for aarch64.

It might be better to just vendor the required code in `deps` so its clear its from another project we might be able to update it in the future.

pgwipeout commented:
> Im not sure if all this code really belongs in `libobs/util`, since its mostly
> https://github.com/nemequ/simde &
> https://github.com/nemequ/portable-snippets
> shredded to fit for aarch64.
> 
> It might be better to just vendor the required code in `deps` so its clear its from another project we might be able to update it in the future.

It is 100% https://github.com/nemequ/simde, and you are correct that it would be better to include it in `deps`.
Unfortunately, the way the build system and install system differ means that including it in `deps` breaks module building after installation.
If I was to include the deps/simde files as inclusions in the headers, their relative position changes once installed, which breaks the include statements.

This is why I pulled just the files used and moved them to util/aarch.
If you have a better way of making cmake include the files in a sane manner, I'll be happy to implement it.

pgwipeout commented:
> > Im not sure if all this code really belongs in `libobs/util`, since its mostly
> > https://github.com/nemequ/simde &
> > https://github.com/nemequ/portable-snippets
> > shredded to fit for aarch64.
> > It might be better to just vendor the required code in `deps` so its clear its from another project we might be able to update it in the future.
> 
> It is 100% https://github.com/nemequ/simde, and you are correct that it would be better to include it in `deps`.
> Unfortunately, the way the build system and install system differ means that including it in `deps` breaks module building after installation.
> If I was to include the deps/simde files as inclusions in the headers, their relative position changes once installed, which breaks the include statements.
> 
> This is why I pulled just the files used and moved them to util/aarch.
> If you have a better way of making cmake include the files in a sane manner, I'll be happy to implement it.

A reasonable solution would be to throw simde into util/simde as a submodule.
Unless there is something else I'm not seeing about the install process that would work.

kkartaltepe commented:
> A reasonable solution would be to throw simde into util/simde as a submodule.

Submodules are already a huge headache so I doubt anything adding more submodules would be be accepted. I dont really feel particularly strongly about the vendoring, but I also dont merge PRs. Thanks for explaining it.

jp9000 commented:
I updated the PR just to reduce unnecessarily repeated code.  Could you check to see if this works?

Additionally, we cannot guarantee that this will be properly maintained.  Anyone using powerpc will have to be responsible for using/maintaining this themselves in the mean time.

pgwipeout commented:
Was it necessary to hijack my copyright notice on the translation header?
Otherwise your changes work fine.

As for maintaining the code, arm64 is very much the field of dreams, as such if you make it build-able, they will come.
I've found plenty of forums asking for this feature.
None of the arm64 boards right now work perfectly out of the box, as such we are used to having to put in work to make things work.

It works well on the Jetson Nano running Ubuntu 19.10.
As such it should work on any modern Tegra chipset (Shield, Switch, etc).
Also, the rk3399 and raspberry pi 4 should run it well, but I haven't personally tested these yet.

pgwipeout commented:
Force committing your changes to my repository was not proper.
Your changes should have been applied as a separate commit.
I have corrected this discrepancy.
Please feel free to amend your commit to correct the copyright header.

jp9000 commented:
I'm sorry, I didn't actually mean to do that, I just copied another header

jp9000 commented:
In the future, I'll use "fixup!" commits rather than that to be safe.  Apologies again for the trouble

pgwipeout commented:
Not a problem.
I apologize for the harsh response, in my line of work one cannot afford to mistake ill intentions.

I will be addressing the concerns brought up by kkartaltepe in one more patch here shortly.

pgwipeout commented:
I've corrected the header issue.
I am deferring the architecture selection fixup to your expertise, since there are several points in the cmake files that check architecture, and I don't want to break anything. (This is a simple feature add after all).

jp9000 commented:
If you can squash any commits that fix up prior commits, then that sounds fine.

pgwipeout commented:
I've rebased to the most recent base and squashed the commit as requested.
Should be ready for your review.

Thanks!
